components/brave_wallet_ui/common/slices/endpoints/nfts.endpoints.ts

```diff
@@ -0,0 +1,249 @@
+      },
+      GetBlockchainTokenIdArg
+    >({
+      queryFn: async (tokenArg, _api, _extraOptions, baseQuery) => {
+        if (!tokenArg.isErc721) {
+          return {
+            error: 'Cannot fetch erc-721 metadata for non erc-721 token'
+          }
+        }
+
+        const { jsonRpcService } = baseQuery(undefined).data
+
+        const result = await jsonRpcService.getERC721Metadata(
+          tokenArg.contractAddress,
+          tokenArg.tokenId,
+          tokenArg.chainId
+        )
+
+        if (result.error || result.errorMessage) {
+          return { error: result.errorMessage }
+        }
+
+        try {
+          const metadata: ERC721Metadata = JSON.parse(result.response)
+          return {
+            data: {
+              id: blockchainTokenEntityAdaptor.selectId(tokenArg),
+              metadata
+            }
+          }
+        } catch (error) {
+          return {
+            error: `error parsing erc721 metadata result: ${result.response}`
+          }
+        }
+      },
+      providesTags: cacher.cacheByBlockchainTokenArg('ERC721Metadata')
+    }),
+    getNftDiscoveryEnabledStatus: query<boolean, void>({
+      queryFn: async (_arg, _api, _extraOptions, _baseQuery) => {
+        try {
+          const { braveWalletService } = apiProxyFetcher()
+          const result = await braveWalletService.getNftDiscoveryEnabled()
+          return {
+            data: result.enabled
+          }
+        } catch (error) {
+          return { error: 'Failed to fetch NFT auto-discovery status' }
+        }
+      },
+      providesTags: ['NftDiscoveryEnabledStatus']
+    }),
+    setNftDiscoveryEnabled: mutation<
+      boolean, // success
+      boolean
+    >({
+      queryFn: async (arg, _api, _extraOptions, _baseQuery) => {
+        try {
+          const { braveWalletService } = apiProxyFetcher()
+          await braveWalletService.setNftDiscoveryEnabled(arg)
+
+          return {
+            data: true
+          }
+        } catch (error) {
+          return {
+            error: 'Failed to set NFT auto-discovery status'
+          }
+        }
+      },
+      invalidatesTags: ['NftDiscoveryEnabledStatus']
+    }),
+    getNftMetadata: query<NFTMetadataReturnType, BraveWallet.BlockchainToken>({
+      queryFn: async (arg, _api, _extraOptions, baseQuery) => {
+        try {
+          const { data: api } = baseQuery(undefined)
+          const { jsonRpcService } = api
+          const result =
+            arg.coin === BraveWallet.CoinType.ETH
+              ? await jsonRpcService.getERC721Metadata(
+                  arg.contractAddress,
+                  arg.tokenId,
+                  arg.chainId
+                )
+              : arg.coin === BraveWallet.CoinType.SOL
+              ? await jsonRpcService.getSolTokenMetadata(
+                  arg.chainId,
+                  arg.contractAddress
+                )
+              : undefined
+
+          if (!result?.error) {
```

Comment:
nit: this might look cleaner if you throw early. that way the bulk of the logic wont be wrapped in the `if`
```suggestion
          if (result?.error) {
            throw new Error(result.error)
          }
```